Dump (some) filelists in the PrettyStackTrace

jrose-apple opened the pull request:
(and later, the derived output file map)

This may help us reproduce a failing build when all we have is a build log, and will become much more important in batch mode when we *really* need to know what ended up in what batch.

Also make a micro-optimization to driver startup time.

jrose-apple commented:
@gottesmm, can you review the first commit; @davidungar and @graydon the second?

@swift-ci Please test

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
Paranoia: things are building slowly on my machine and I want to make sure it's not my change:

@swift-ci Please smoke test compiler performance

jrose-apple commented:
@swift-ci Please smoke test compiler performance

gottesmm commented on include/swift/Basic/LLVMInitialize.h:
```diff
@@ -10,34 +10,31 @@
-  llvm::llvm_shutdown_obj _INITIALIZE_LLVM_SHUTDOWN_OBJ; \
+  llvm::llvm_shutdown_obj _INITIALIZE_LLVM_SHUTDOWN_OBJ
+
+#define INITIALIZE_LLVM() \
```
Since INITIALIZE_LLVM doesn't care about scope anymore, can you wrap the body in a do while(0)?

gottesmm reviewed (commented):
Great idea! One quick comment, but otherwise looks good to me.

davidungar reviewed (commented):
If I understand the code, it's passing a bunch of buffers down into the input conversion object in order to be able to print out the filelist contents on a crash.

I'm not sure this is the best way to go about dumping the information you want. If you want the filelist contents, I would open up the interface to the appropriate classes to get those data specifically, rather than passing anonymous buffers around. The issue I see with this formulation is that one would not know which buffer holds which filelist without deciphering the code.

I also worry that the copy-optimization by passing down buffers adds fragility without saving significant time. It spreads the ownership invariants across classes, rather than confining them to one class. In general, we do optimize out copying, but in this case I'm not sure it's worth the added fragility.

The fallback buffer scheme seems like it might also be a potential source of bugs down the line.

My preferred alternative would be let the paths computers return the desired filelist contents, and to copy them into the FrontendInputsAndOutputs, where the stack-trace-dumping code in FrontendendTool.cpp could find them.

graydon reviewed (approved):
Looks highly reasonable to me!

davidungar commented:
It wasn't clear to me from reading the code that having the calling code not know about the contents of the buffers was so deliberate. A Doxygen comment (or did I miss one?) for those buffers might be helpful. For example after:
      /// Any configuration files loaded as a result of parsing arguments will be
     /// stored in \p ConfigurationFileBuffers, if non-null.
You could add something like: "the contents of these buffers is deliberately opaque because they will only be used to help reproduce crashes when the user is familiar with the code anyway."

Hmmm, what about the case where the compiler crashes for a developer who is not a compiler whiz. Might it not be helpful to print out a bit more about buffer contents? But I guess that's out-of-scope for what you want to do here. When I was writing apps, I would get compiler crashes and try to figure out a workaround from the stack trace. That's why I thought more printed into might help. But, OK, out of scope here.

I continue to favor the out-parameter approach, because it reduces the coupling between the parts of the compiler. I don't think it would make a difference to you because you know this code base through-and-through, but to noobs to the compiler such as I am, it's been very easy to introduce ownership bugs--I've been spending over a day on the latest one--and I think that the out-parameter approach makes the lifetimes and causality flow much clearer.

davidungar commented:
Why is `BuffersFallback` needed? Under what circumstances would there be no configuration buffers?
Would it help to add a comment there?

jrose-apple commented:
I'd argue that passing a SmallVector out through a pointer is really *more* complicated than passing a vector in that gets filled in. That's a pretty common idiom across the codebase, although admittedly a bunch of that predates C++11 mode and move constructors.

`BuffersFallback` is there to allow callers who don't plan to do anything with the buffers to not have to pass an argument at all. There are several such callers today (SourceKit, all of the SIL tools, probably LLDB).

jrose-apple committed:
Dump input filelists in the PrettyStackTrace

This may help us reproduce a failing build when all we have is a build
log, and will become much more important in batch mode when we
/really/ need to know what ended up in what batch.

For now, this doesn't include /output/ filelists, because David's
about to mess with that code anyway to make things better around
supplementary outputs in batch mode. There is one weirdness there,
though, which is that ArgsToFrontendInputsConverter peeks at the
outputs to see whether we're doing single-threaded or multi-threaded
WMO.

jrose-apple commented:
Here's a version that is more explicit about its out-parameter usage. What do you think?

@swift-ci Please smoke test

davidungar commented:
I like the new version a lot more! I have a few thoughts, some about this code, and some about working in this world in general for my own benefit in learning how to do so.

Regarding this PR:
I like the `takeConfigurationFileBuffers` a lot because it really says to me what's going on. I thought about merging it in with convert. What I mean is passing the `ConfigurationFileBuffers` in as an out parameter to the various `convert` functions. Then, the `convert` function in `ArgsToFrontendOptionsConverter` could in turn pass the parameter down to `ArgsToFrontendInputsConverter::convert`, to `computeMainAndSupplementaryOutputFilenames`, and that in turn could pass the out parameter down to `ArgsToFrontendOutputsConverter::convert`.
I like the idea of calling the take routine inside of convert, although it does broaden the scope of what we mean by "convert", because I'm fine with adding the breadcrumb dropping for stack tracing to the notion of conversion. I like the integration with convert because whenever we convert, we also want to drop the breadcrumb.

If I read the code correctly, it is not (yet?) capturing all the file lists, merely the input file lists. Would it be good to clarify the comment “Any configuration files loaded as a result of parsing arguments will be"?

Here's perhaps a broader style question: When do we want to use a pointer that may be nil, as opposed to a reference to an Optional? The reference to an Optional seems more explicit to me--I love what Swift does with optionals--but I don't understand all the ramifications here. If the pointer is better, I assume there's a comment (or should be) somewhere that says the pointer may be null if it should be ignored.

Thanks!

davidungar commented on include/swift/Frontend/ArgsToFrontendInputsConverter.h:
```diff
@@ -47,7 +47,9 @@ class ArgsToFrontendInputsConverter {
-  SmallVector<std::unique_ptr<llvm::MemoryBuffer>, 4> BuffersToKeepAlive;
+  /// A place to keep alive any buffers that are loaded as part of setting up
+  /// the frontend inputs.
+  SmallVector<std::unique_ptr<llvm::MemoryBuffer>, 4> ConfigurationFileBuffers;
```
Nice!

davidungar commented on include/swift/Frontend/Frontend.h:
```diff
@@ -86,11 +86,19 @@ class CompilerInvocation {
+  /// buffers should \e not be interpreted by the caller; they are only present
+  /// in order to make it possible to reproduce how these arguments were parsed
+  /// if the compiler ends up crashing or exhibiting other bad behavior.
+  ///
```
Nice!

davidungar commented on lib/Frontend/ArgsToFrontendInputsConverter.cpp:
```diff
@@ -36,6 +36,13 @@ ArgsToFrontendInputsConverter::ArgsToFrontendInputsConverter(
+  std::move(ConfigurationFileBuffers.begin(), ConfigurationFileBuffers.end(),
+            std::back_inserter(buffers));
+  ConfigurationFileBuffers.clear();
+}
```
Here's something I don't know about these data structures: if you move stuff out of ConfigurationFileBuffers, why do you also need to clear them afterwards?

jrose-apple replied:
`std::move` works on an arbitrary range, so it can't actually resize the container itself. It leaves behind null `std::unique_ptr`s instead. The `clear` isn't strictly necessary since we throw away the converter immediately after this, but if we *were* to use it for something else we wouldn't want to have a bunch of nulls lying around.

davidungar replied:
Thanks!

davidungar commented on test/Frontend/crash.swift:
```diff
@@ -0,0 +1,13 @@
+// RUN: echo %s > %t.filelist.txt
```
I know that the following isn't our typical practice, but as a noob, it would help me to add a one sentence comment to this test describing what it does.

jrose-apple committed:
Reorganize ArgsToFrontendInputsConverter to be more functional-style

Per a conversation with DavidU.

jrose-apple commented:
David and I talked about this one offline, then I went to go try things out, and now I'm in a quandry because different parts of the code are suggesting different ways to pass the file buffers up the stack and it's bothering me a lot. David, how much do you care about the buffers being returned functionally? It interfaces *really terribly* with the non-functional parts because of the unique_ptrs being move-only.

davidungar commented:
If I weren’t in class, I’d come see you in person. I’m intrigued by what you wrote. But sure, I’m very happy to follow your judgement. Whichever way you have decided is fine with me. 

- David

> On Feb 28, 2018, at 3:41 PM, Jordan Rose <notifications@github.com> wrote:
> 
> David and I talked about this one offline, then I went to go try things out, and now I'm in a quandry because different parts of the code are suggesting different ways to pass the file buffers up the stack and it's bothering me a lot. David, how much do you care about the buffers being returned functionally? It interfaces really terribly with the non-functional parts because of the unique_ptrs being move-only.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

jrose-apple commented:
@davidungar Came back to this and put in some of the refactoring we talked about a while back. I also decided to just do the outputs later so that we can at least get this part in.

@swift-ci Please test

davidungar commented on lib/FrontendTool/FrontendTool.cpp:
```diff
@@ -1626,10 +1627,29 @@ int swift::performFrontend(ArrayRef<const char *> Args,
+              Optional<PrettyStackTraceFileContents> &trace) {
+    trace.emplace(*buffer);
+  });
+
```
I would be tempted to enclose 1629-1651 in curly braces to show the reader that `configurationFileBuffers` and `configurationFileStackTraces` were not used again in this function.

jrose-apple replied:
Ah, yeah, I'd love to…but they actually need to stay alive, since the PrettyStackTrace implementation is a linked list with stack-based lifetimes.

davidungar replied:
Did I overlook a reference to it down below? Or is it non-obvious?? If the latter, you could save me some potential grief in the future with a comment to that effect in the code. Otherwise, I would come along, add the braces, and mess things up!

jrose-apple replied:
It's RAII. I think we don't bother usually because "PrettyStackTrace" is already something weird, but in this case that type is buried in the unique_ptr, so I can make the comment more verbose.

davidungar replied:
Yes. I have a habit of grepping for a local identifier and isolating it in a block if I don't see it used later. Since I tend to skim, it's easy for me to miss the fact than something is RAII and needs to stay alive, perhaps also because we don't always isolate locals in blocks in long routines when they could be.

davidungar commented on lib/Frontend/ArgsToFrontendInputsConverter.cpp:
```diff
@@ -30,33 +31,48 @@ using namespace swift;
+  std::tie(result, unusedPrimaryFiles) =
       createInputFilesConsumingPrimaries(*primaryFiles);
 
   if (checkForMissingPrimaryFiles(unusedPrimaryFiles))
```
This one is my own fault, but I think the use of both "missing" and "unused" in `unusedPrimaryFiles` and `checkForMissingPrimaryFiles` could confuse a reader. Would you be interested in coming up with a better, clearer, more consistent nomenclature for these?

jrose-apple replied:
I canonicalized on "unused", for a local improvement. (I guess I could have left this alone too since I hadn't previously touched it in this PR, but it seems okay.)

davidungar commented on lib/Frontend/ArgsToFrontendInputsConverter.cpp:
```diff
@@ -30,33 +31,48 @@ using namespace swift;
-  InputsAndOutputs.setIsSingleThreadedWMO(isSingleThreadedWMO());
-
-  InputsAndOutputs.setBypassBatchModeChecks(
+  result.setIsSingleThreadedWMO(isSingleThreadedWMO(result));
```
There's something funny here; the line reads oddly to my eye, as if it's checking the result and then setting the same thing it just checked for. `result` is being both checked and set, and `isSingleThreadedWMO` looks like it could be a static function here, too.
As I look at it, I think that this functionality might work better if moved to  the end of`createInputFilesConsumingPrimaries`, because that function creates and returns a `FrontendInputsAndOutputs`, so `createInputFilesConsumingPrimaries` can look at its state and do the computation.

Alternatively, this functionality could be moved into `FrontendInputsAndOutputs` as no-argument `cacheIsSingleThreadedWMO` function.

jrose-apple replied:
The `createInputFilesConsumingPrimaries` worked!

davidungar reviewed (approved):
This is beautiful! I love it. It's certainly merge quality. I have added a few thoughts on little tweaks for your consideration.

jrose-apple commented:
@swift-ci Please smoke test

davidungar commented:
Looks fine!

- David

> On Mar 9, 2018, at 6:25 PM, Jordan Rose <notifications@github.com> wrote:
> 
> @jrose-apple commented on this pull request.
> 
> In lib/Frontend/ArgsToFrontendInputsConverter.cpp:
> 
> >    Optional<std::set<StringRef>> primaryFiles = readPrimaryFiles();
>    if (!primaryFiles)
> -    return true;
> -  std::set<StringRef> unusedPrimaryFiles =
> +    return None;
> +
> +  FrontendInputsAndOutputs result;
> +  std::set<StringRef> unusedPrimaryFiles;
> +  std::tie(result, unusedPrimaryFiles) =
>        createInputFilesConsumingPrimaries(*primaryFiles);
>  
>    if (checkForMissingPrimaryFiles(unusedPrimaryFiles))
> I canonicalized on "unused", for a local improvement. (I guess I could have left this alone too since I hadn't previously touched it in this PR, but it seems okay.)
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

davidungar commented on lib/Frontend/ArgsToFrontendInputsConverter.cpp:
```diff
@@ -137,34 +152,38 @@ ArgsToFrontendInputsConverter::readPrimaryFiles() {
+        OutputFilesComputer::getOutputFilenamesFromCommandLineOrFilelist(Args,
+                                                                         Diags);
+    if (userSuppliedNamesOrErr && userSuppliedNamesOrErr->size() == 1)
+      result.setIsSingleThreadedWMO(true);
```
Cool!

davidungar reviewed (approved):
This came out very nice!

davidungar commented:
Nice comment!

- David

> On Mar 8, 2018, at 8:15 PM, Jordan Rose <notifications@github.com> wrote:
> 
> @jrose-apple commented on this pull request.
> 
> In lib/FrontendTool/FrontendTool.cpp:
> 
> > +  // we used to parse the arguments.
> +  //
> +  // This can't be a SmallVector or similar because PrettyStackTraces can't be
> +  // moved (or copied)...and it can't be an array of non-optionals because
> +  // PrettyStackTraces can't be default-constructed. So we end up with a
> +  // dynamically-sized array of optional PrettyStackTraces, which get
> +  // initialized by iterating over the buffers we collected above.
> +  auto configurationFileStackTraces = llvm::make_unique<Optional<PrettyStackTraceFileContents>[]>(
> +        configurationFileBuffers.size());
> +  for_each(configurationFileBuffers.begin(), configurationFileBuffers.end(),
> +           &configurationFileStackTraces[0],
> +           [](const std::unique_ptr<llvm::MemoryBuffer> &buffer,
> +              Optional<PrettyStackTraceFileContents> &trace) {
> +    trace.emplace(*buffer);
> +  });
> +
> It's RAII. I think we don't bother usually because "PrettyStackTrace" is already something weird, but in this case that type is buried in the unique_ptr, so I can make the comment more verbose.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>
